src/uct/ib/mlx5/ib_mlx5.c

```diff
@@ -543,48 +543,108 @@ int uct_ib_mlx5_devx_uar_cmp(uct_ib_mlx5_devx_uar_t *uar,
-uct_ib_mlx5_devx_alloc_uar(uct_ib_mlx5_md_t *md, unsigned flags, int log_level,
-                           char *title, char *fallback,
-                           struct mlx5dv_devx_uar **uar_p)
+static ucs_status_t uct_ib_mlx5_devx_alloc_uar(uct_ib_mlx5_md_t *md, 
+                                               uint32_t flags,
+                                               struct mlx5dv_devx_uar **uar_p) 
 {
+    const char *pf_log_bar_size_suggestion = "Consider increasing PF_LOG_BAR_SIZE using mlxconfig tool (requires reboot)";
+    const char *allocation_type            = (flags == UCT_IB_MLX5_UAR_ALLOC_TYPE_WC) ? "WC" : "NC_DEDICATED";
+    ucs_log_level_t err_log_level          = UCS_LOG_LEVEL_DIAG;
     struct mlx5dv_devx_uar *uar;
-    char buf[512];
+    ucs_status_t status;
+    ucs_string_buffer_t buf;
+    int err;
 
     uar = mlx5dv_devx_alloc_uar(md->super.dev.ibv_context, flags);
     if (uar == NULL) {
-        sprintf(buf, "mlx5dv_devx_alloc_uar(device=%s, flags=0x%x(%s)) "
-                "failed: %m", uct_ib_device_name(&md->super.dev), flags, title);
-        if (fallback == NULL) {
-            ucs_log(log_level, "%s", buf);
-        } else {
-            ucs_log(log_level, "%s, fallback to %s", buf, fallback);
+        ucs_string_buffer_init(&buf);
+
+        err = errno;
+        ucs_string_buffer_appendf(&buf, "mlx5dv_devx_alloc_uar(device=%s, flags=0x%x) type=%s failed: %s. ",
+                                  uct_ib_device_name(&md->super.dev), flags, allocation_type, strerror(err));
+        
+        switch (err) {
+            case ENOMEM:
+                /* UAR allocation failed due to insufficient PF_LOG_BAR_SIZE; fallback will produce the same result */
+                ucs_string_buffer_appendf(&buf, "%s", pf_log_bar_size_suggestion);
+                err_log_level = UCS_LOG_LEVEL_ERROR;
+                status        = UCS_ERR_NO_MEMORY;
+                break;
+
+            case EOPNOTSUPP:
+                /* UAR allocation failed due to unsupported allocation type, if there's no fallback, it's an error */
+                status = UCS_ERR_UNSUPPORTED;
+                break;
+
+            case EINVAL:
+                status = UCS_ERR_INVALID_PARAM;
+                break;
+
+            default:
+                status = UCS_ERR_NO_MEMORY;
+                break;
         }
-
-        return UCS_ERR_NO_MEMORY;
+        ucs_log(err_log_level, "%s", ucs_string_buffer_cstr(&buf));
+        ucs_string_buffer_cleanup(&buf);
+        return status;
     }
 
     *uar_p = uar;
     return UCS_OK;
 }
 #endif
 
+ucs_status_t uct_ib_mlx5_devx_check_supported_uar_alloc_type(uct_ib_mlx5_md_t *md)
+{
+#if HAVE_DEVX
+    uct_ib_mlx5_devx_uar_t uar;
+    ucs_status_t status;
+
+    status = uct_ib_mlx5_devx_alloc_uar(md, UCT_IB_MLX5_UAR_ALLOC_TYPE_WC, &uar.uar);
+    if (status == UCS_ERR_UNSUPPORTED) {
+        ucs_debug("WC UAR allocation type is not supported, falling back to NC_DEDICATED");
```

Comment:
1. print device name in all log messages in this function
2. "falling back to NC_DEDICATED" is misleading because it also may be unsupported. IMO better just print the failure and not the future action
3. uct_ib_mlx5_devx_alloc_uar already prints a lot of log messages and has logic to print the PF_LOG_BAR_SIZE suggestion. I think here we need to use mlx5dv_devx_alloc_uar directly

Reply:
1. done
2. removed it
3. IMO the entire logic of uct_ib_mlx5_devx_alloc_uar should be used here, because mlx5dv_devx_uar can fail for different reasons, it can reach ENOMEM here as well.